src/steth/scUSDCv2.sol

```diff
@@ -0,0 +1,580 @@
+
+    event EmergencyExitExecuted(
+        address indexed admin, uint256 wethWithdrawn, uint256 debtRepaid, uint256 collateralReleased
+    );
+    event Reallocated();
+    event Rebalanced(uint256 totalCollateral, uint256 totalDebt, uint256 floatBalance);
+    event ProfitSold(uint256 wethSold, uint256 usdcReceived);
+    event EulerRewardsSold(uint256 eulerSold, uint256 usdcReceived);
+
+    // token representing the rewards from the euler protocol
+    ERC20 public constant eulerRewardsToken = ERC20(C.EULER_REWARDS_TOKEN);
+
+    // Uniswap V3 router
+    ISwapRouter public immutable swapRouter;
+
+    // Chainlink pricefeed (USDC -> WETH)
+    AggregatorV3Interface public usdcToEthPriceFeed;
+
+    // Balancer vault for flashloans
+    IVault public immutable balancerVault;
+
+    // mapping of protocol IDs to adapters
+    EnumerableMap.UintToAddressMap private protocolAdapters;
+
+    struct ConstructorParams {
+        address admin;
+        address keeper;
+        ERC4626 scWETH;
+        ERC20 usdc;
+        WETH weth;
+        ISwapRouter uniswapSwapRouter;
+        AggregatorV3Interface chainlinkUsdcToEthPriceFeed;
+        IVault balancerVault;
+    }
+
+    constructor(ConstructorParams memory _params)
+        scUSDCBase(
+            _params.admin,
+            _params.keeper,
+            _params.usdc,
+            _params.weth,
+            _params.scWETH,
+            "Sandclock USDC Vault v2",
+            "scUSDCv2"
+        )
+    {
+        swapRouter = _params.uniswapSwapRouter;
+        usdcToEthPriceFeed = _params.chainlinkUsdcToEthPriceFeed;
+        balancerVault = _params.balancerVault;
+
+        weth.safeApprove(address(swapRouter), type(uint256).max);
+        weth.safeApprove(address(scWETH), type(uint256).max);
+    }
+
+    /*//////////////////////////////////////////////////////////////
+                            PUBLIC API
+    //////////////////////////////////////////////////////////////*/
+
+    /**
+     * @notice Set the chainlink price feed for USDC -> WETH.
+     * @param _newPriceFeed The new price feed.
+     */
+    function setUsdcToEthPriceFeed(AggregatorV3Interface _newPriceFeed) external {
+        _onlyAdmin();
+
+        if (address(_newPriceFeed) == address(0)) revert PriceFeedZeroAddress();
+
+        usdcToEthPriceFeed = _newPriceFeed;
+    }
+
+    function addAdapter(IAdapter _adapter) external {
+        _onlyAdmin();
+
+        uint8 id = _adapter.id();
+
+        if (isSupported(id)) revert ProtocolInUse(id);
+
+        protocolAdapters.set(uint256(id), address(_adapter));
+
+        address(_adapter).functionDelegateCall(abi.encodeWithSelector(IAdapter.setApprovals.selector));
+    }
+
+    function removeAdapter(uint8 _adapterId) external {
+        _onlyAdmin();
+        _isSupportedCheck(_adapterId);
+
+        // check if protocol is being used
+        if (IAdapter(protocolAdapters.get(_adapterId)).getCollateral(address(this)) > 0) {
```

Comment:
You should probably have checkForFunds boolean here. Because if the getCollateral method reverts (like in the case of Euler after the hack) we won't be able to remove the adapter since it will revert here everytime.

Reply:
got it, will add

Reply:
[fixed](https://github.com/lindy-labs/sandclock-contracts/pull/58/commits/97cc7920aae5d4a9be4b3f398172dddcab3ae7c6)